### PR TITLE
fix($core): register async page/layout components and pascalize their names (close: #1173, #1321, #1391) [WIP]

### PR DESCRIPTION
**Pull request by Shigma**
**Summary**

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [x] Code style update
- [x] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI of default theme, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

**Comment by Shigma** [comment]
cc @meteorlxy

File: packages/@vuepress/core/lib/client/util.js [hunk 1]
```diff
@@ -4,6 +4,11 @@ import pageComponents from '@internal/page-components'
 
 const asyncComponents = Object.assign({}, layoutComponents, pageComponents)
 
+// TODO: reuse this function in shared-utils
+function pascalize (source = '') {
```
Comment (meteorlxy) [inline packages/@vuepress/core/lib/client/util.js]: Consider use a `cached` fn like https://github.com/vuejs/vue/blob/731e4d0ebc997a6c2b1fc500b08924ff7ba091b6/src/shared/util.js#L153-L159

Comment (Shigma) [reply packages/@vuepress/core/lib/client/util.js]: Considering that this function might come from shared-utils in the future, this sort of optimization will be implemented after then.

**Comment by ulivz** [comment]
Closing it in favor of 9d0151483b330f7931d46ed9ad4c501e7e47d289, 0306574b1f852e3284a65c62d2c85a1e0fefe97a and 3e91eba5beed4e94367ae6b993b02b3bb89976ae

**Pull request closed without merging**